File: python/polars_redis/_multi.py
"""Multi-cluster write support for polars-redis.

This module provides functions for writing DataFrames to multiple Redis
clusters in parallel, enabling fan-out patterns for replication and
geographic distribution.
"""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass, field
from typing import Any, Callable

import polars as pl


@dataclass
class Destination:
    """Configuration for a write destination.

    Attributes:
        url: Redis connection URL (e.g., "redis://localhost:6379").
        name: Optional name for this destination (for logging/metrics).
        key_prefix: Prefix to prepend to all keys (default: "").
        transform: Optional function to transform the DataFrame before writing.
        timeout_ms: Timeout for this destination in milliseconds (default: 30000).
    """

    url: str
    name: str | None = None
    key_prefix: str = ""
    transform: Callable[[pl.DataFrame], pl.DataFrame] | None = None
    timeout_ms: int = 30000

    def __post_init__(self) -> None:
        if self.name is None:
            # Extract host from URL for default name
            self.name = self.url.split("://")[-1].split("/")[0]


@dataclass
class DestinationResult:
    """Result of a write operation to a single destination.

    Attributes:
        destination: The destination configuration.
        keys_written: Number of keys successfully written.
        keys_failed: Number of keys that failed to write.
        success: Whether the write completed without errors.
        error: Error message if the write failed entirely.
        duration_ms: Time taken for the write in milliseconds.
    """

    destination: Destination
    keys_written: int = 0
    keys_failed: int = 0
    success: bool = True
    error: str | None = None
    duration_ms: float = 0.0

    def __repr__(self) -> str:
        status = "ok" if self.success else f"error: {self.error}"
        return (
            f"DestinationResult({self.destination.name}: "
            f"{self.keys_written} written, {self.keys_failed} failed, {status})"
        )


@dataclass
class MultiWriteResult:
    """Aggregated result of writing to multiple destinations.

    Attributes:
        results: List of per-destination results.
        total_keys_written: Total keys written across all destinations.
        total_keys_failed: Total keys failed across all destinations.
        all_succeeded: Whether all destinations completed successfully.
        failed_destinations: List of destination names that failed.
    """

    results: list[DestinationResult] = field(default_factory=list)

    @property
    def total_keys_written(self) -> int:
        return sum(r.keys_written for r in self.results)

    @property
    def total_keys_failed(self) -> int:
        return sum(r.keys_failed for r in self.results)

    @property
    def all_succeeded(self) -> bool:
        return all(r.success for r in self.results)

    @property
    def failed_destinations(self) -> list[str]:
        return [r.destination.name or r.destination.url for r in self.results if not r.success]

    def __repr__(self) -> str:
        succeeded = sum(1 for r in self.results if r.success)
        failed = len(self.results) - succeeded
        return (
            f"MultiWriteResult({succeeded}/{len(self.results)} destinations succeeded, "
            f"{self.total_keys_written} total keys written)"
        )


async def _write_to_destination_hashes(
    df: pl.DataFrame,
    dest: Destination,
    key_column: str | None,
    ttl: int | None,
    if_exists: str,
) -> DestinationResult:
    """Write DataFrame to a single destination as hashes."""
    import time

    from polars_redis._internal import py_write_hashes

    start = time.perf_counter()
    result = DestinationResult(destination=dest)

    try:
        # Apply transform if provided
        write_df = dest.transform(df) if dest.transform else df

        # Prepare keys and values
        if key_column is None:
            keys = [f"{dest.key_prefix}{i}" for i in range(len(write_df))]
            field_columns = list(write_df.columns)
        else:
            if key_column not in write_df.columns:
                raise ValueError(f"Key column '{key_column}' not found in DataFrame")
            keys = [f"{dest.key_prefix}{k}" for k in write_df[key_column].to_list()]
            field_columns = [c for c in write_df.columns if c != key_column]

        # Convert values to strings
        values = []
        for i in range(len(write_df)):
            row_values = []
            for col in field_columns:
                val = write_df[col][i]
                row_values.append(None if val is None else str(val))
            values.append(row_values)

        # Run the sync write in a thread pool to not block the event loop
        loop = asyncio.get_event_loop()
        keys_written, keys_failed, _ = await asyncio.wait_for(
            loop.run_in_executor(
                None,
                lambda: py_write_hashes(dest.url, keys, field_columns, values, ttl, if_exists),
            ),
            timeout=dest.timeout_ms / 1000,
        )

        result.keys_written = keys_written
        result.keys_failed = keys_failed
        result.success = True

    except asyncio.TimeoutError:
        result.success = False
        result.error = f"Timeout after {dest.timeout_ms}ms"
    except Exception as e:
        result.success = False
        result.error = str(e)

    result.duration_ms = (time.perf_counter() - start) * 1000
    return result


async def _write_to_destination_json(
    df: pl.DataFrame,
    dest: Destination,
    key_column: str | None,
    ttl: int | None,
    if_exists: str,
) -> DestinationResult:
    """Write DataFrame to a single destination as JSON."""
    import time

    from polars_redis._internal import py_write_json

    start = time.perf_counter()
    result = DestinationResult(destination=dest)

    try:
        # Apply transform if provided
        write_df = dest.transform(df) if dest.transform else df

        # Prepare keys and JSON strings
        if key_column is None:
            keys = [f"{dest.key_prefix}{i}" for i in range(len(write_df))]
            field_columns = list(write_df.columns)
        else:
            if key_column not in write_df.columns:
                raise ValueError(f"Key column '{key_column}' not found in DataFrame")
            keys = [f"{dest.key_prefix}{k}" for k in write_df[key_column].to_list()]
            field_columns = [c for c in write_df.columns if c != key_column]

        # Build JSON strings
        json_strings = []
        for i in range(len(write_df)):
            doc = {}
            for col in field_columns:
                val = write_df[col][i]
                if val is not None:
                    doc[col] = val
            json_strings.append(json.dumps(doc))

        # Run the sync write in a thread pool
        loop = asyncio.get_event_loop()
        keys_written, keys_failed, _ = await asyncio.wait_for(
            loop.run_in_executor(
                None,
                lambda: py_write_json(dest.url, keys, json_strings, ttl, if_exists),
            ),
            timeout=dest.timeout_ms / 1000,
        )

        result.keys_written = keys_written
        result.keys_failed = keys_failed
        result.success = True

    except asyncio.TimeoutError:
        result.success = False
        result.error = f"Timeout after {dest.timeout_ms}ms"
    except Exception as e:
        result.success = False
        result.error = str(e)

    result.duration_ms = (time.perf_counter() - start) * 1000
    return result


async def write_hashes_multi(
    df: pl.DataFrame,
    destinations: list[Destination | dict[str, Any]],
    key_column: str | None = "_key",
    ttl: int | None = None,
    if_exists: str = "replace",
) -> MultiWriteResult:
    """Write a DataFrame to multiple Redis clusters as hashes in parallel.

    Each destination receives the same DataFrame (optionally transformed),
    written as Redis hashes. Writes execute concurrently, and failures at
    one destination don't affect others.

    Args:
        df: The DataFrame to write.
        destinations: List of destinations. Each can be a Destination object
            or a dict with keys: url (required), name, key_prefix, transform, timeout_ms.
        key_column: Column containing Redis keys (default: "_key").
            If None, keys are auto-generated from row indices.
        ttl: Optional TTL in seconds for each key.
        if_exists: How to handle existing keys ("fail", "replace", "append").

    Returns:
        MultiWriteResult with per-destination results.

    Example:
        >>> import polars as pl
        >>> import polars_redis as pr
        >>> from polars_redis import Destination
        >>>
        >>> df = pl.DataFrame({
        ...     "id": ["user:1", "user:2"],
        ...     "name": ["Alice", "Bob"],
        ...     "region": ["US", "EU"],
        ... })
        >>>
        >>> result = await pr.write_hashes_multi(
        ...     df,
        ...     destinations=[
        ...         Destination(url="redis://cluster-a:6379", key_prefix="primary:"),
        ...         Destination(url="redis://cluster-b:6379", key_prefix="backup:"),
        ...     ],
        ...     key_column="id",
        ... )
        >>>
        >>> print(f"Wrote to {len(result.results)} destinations")
        >>> for r in result.results:
        ...     print(f"  {r.destination.name}: {r.keys_written} keys")
    """
    # Normalize destinations to Destination objects
    normalized: list[Destination] = []
    for dest in destinations:
        if isinstance(dest, dict):
            normalized.append(Destination(**dest))
        else:
            normalized.append(dest)

    # Write to all destinations in parallel
    tasks = [
        _write_to_destination_hashes(df, dest, key_column, ttl, if_exists) for dest in normalized
    ]
    results = await asyncio.gather(*tasks, return_exceptions=False)

    return MultiWriteResult(results=list(results))


async def write_json_multi(
    df: pl.DataFrame,
    destinations: list[Destination | dict[str, Any]],
    key_column: str | None = "_key",
    ttl: int | None = None,
    if_exists: str = "replace",
) -> MultiWriteResult:
    """Write a DataFrame to multiple Redis clusters as JSON documents in parallel.

    Each destination receives the same DataFrame (optionally transformed),
    written as RedisJSON documents. Writes execute concurrently, and failures
    at one destination don't affect others.

    Args:
        df: The DataFrame to write.
        destinations: List of destinations. Each can be a Destination object
            or a dict with keys: url (required), name, key_prefix, transform, timeout_ms.
        key_column: Column containing Redis keys (default: "_key").
            If None, keys are auto-generated from row indices.
        ttl: Optional TTL in seconds for each key.
        if_exists: How to handle existing keys ("fail", "replace", "append").

    Returns:
        MultiWriteResult with per-destination results.

    Example:
        >>> import polars as pl
        >>> import polars_redis as pr
        >>> from polars_redis import Destination
        >>>
        >>> df = pl.DataFrame({
        ...     "id": ["doc:1", "doc:2"],
        ...     "title": ["Hello", "World"],
        ...     "views": [100, 200],
        ... })
        >>>
        >>> result = await pr.write_json_multi(
        ...     df,
        ...     destinations=[
        ...         Destination(url="redis://primary:6379"),
        ...         Destination(url="redis://replica:6379"),
        ...     ],
        ...     key_column="id",
        ... )
        >>>
        >>> if not result.all_succeeded:
        ...     print(f"Failed destinations: {result.failed_destinations}")
    """
    # Normalize destinations to Destination objects
    normalized: list[Destination] = []
    for dest in destinations:
        if isinstance(dest, dict):
            normalized.append(Destination(**dest))
        else:
            normalized.append(dest)

    # Write to all destinations in parallel
    tasks = [
        _write_to_destination_json(df, dest, key_column, ttl, if_exists) for dest in normalized
    ]
    results = await asyncio.gather(*tasks, return_exceptions=False)

    return MultiWriteResult(results=list(results))

File: src/io/multi.rs
//! Multi-cluster write operations.
//!
//! This module provides functionality for writing data to multiple Redis
//! clusters or instances in parallel.
//!
//! # Example
//!
//! ```ignore
//! use polars_redis::{MultiClusterWriter, Destination, WriteMode};
//!
//! let writer = MultiClusterWriter::new(vec![
//!     Destination::new("redis://cluster1:6379").with_name("primary"),
//!     Destination::new("redis://cluster2:6379").with_name("secondary"),
//! ]);
//!
//! let results = writer.write_hashes(
//!     keys,
//!     fields,
//!     values,
//!     Some(3600),
//!     WriteMode::Replace,
//! )?;
//!
//! for result in results {
//!     println!("{}: {} keys written", result.destination_name, result.keys_written);
//! }
//! ```

use std::time::{Duration, Instant};

use tokio::runtime::Runtime;

use crate::error::{Error, Result};
use crate::io::write::{WriteMode, WriteResult, write_hashes, write_json};

/// Configuration for a write destination.
#[derive(Debug, Clone)]
pub struct Destination {
    /// Redis connection URL.
    pub url: String,
    /// Optional name for this destination.
    pub name: Option<String>,
    /// Key prefix to prepend to all keys.
    pub key_prefix: String,
    /// Timeout in milliseconds for write operations.
    pub timeout_ms: u64,
}

impl Destination {
    /// Create a new destination with the given URL.
    pub fn new(url: &str) -> Self {
        Self {
            url: url.to_string(),
            name: None,
            key_prefix: String::new(),
            timeout_ms: 30000,
        }
    }

    /// Set the destination name.
    pub fn with_name(mut self, name: &str) -> Self {
        self.name = Some(name.to_string());
        self
    }

    /// Set the key prefix.
    pub fn with_key_prefix(mut self, prefix: &str) -> Self {
        self.key_prefix = prefix.to_string();
        self
    }

    /// Set the timeout in milliseconds.
    pub fn with_timeout_ms(mut self, ms: u64) -> Self {
        self.timeout_ms = ms;
        self
    }

    /// Get the display name for this destination.
    pub fn display_name(&self) -> String {
        self.name.clone().unwrap_or_else(|| self.url.clone())
    }

    /// Apply the key prefix to a key.
    pub fn prefix_key(&self, key: &str) -> String {
        if self.key_prefix.is_empty() {
            key.to_string()
        } else {
            format!("{}{}", self.key_prefix, key)
        }
    }
}

/// Result of writing to a single destination.
#[derive(Debug, Clone)]
pub struct DestinationResult {
    /// Name of the destination.
    pub destination_name: String,
    /// URL of the destination.
    pub destination_url: String,
    /// Number of keys successfully written.
    pub keys_written: usize,
    /// Number of keys that failed to write.
    pub keys_failed: usize,
    /// Whether the write was successful overall.
    pub success: bool,
    /// Error message if the write failed.
    pub error: Option<String>,
    /// Duration of the write operation in milliseconds.
    pub duration_ms: f64,
}

impl DestinationResult {
    /// Create a successful result.
    pub fn success(dest: &Destination, write_result: &WriteResult, duration: Duration) -> Self {
        Self {
            destination_name: dest.display_name(),
            destination_url: dest.url.clone(),
            keys_written: write_result.keys_written,
            keys_failed: write_result.keys_failed,
            success: true,
            error: None,
            duration_ms: duration.as_secs_f64() * 1000.0,
        }
    }

    /// Create a failed result.
    pub fn failure(dest: &Destination, error: String, duration: Duration) -> Self {
        Self {
            destination_name: dest.display_name(),
            destination_url: dest.url.clone(),
            keys_written: 0,
            keys_failed: 0,
            success: false,
            error: Some(error),
            duration_ms: duration.as_secs_f64() * 1000.0,
        }
    }
}

/// Result of a multi-cluster write operation.
#[derive(Debug, Clone)]
pub struct MultiWriteResult {
    /// Results for each destination.
    pub results: Vec<DestinationResult>,
    /// Total keys written across all destinations.
    pub total_keys_written: usize,
    /// Total keys failed across all destinations.
    pub total_keys_failed: usize,
    /// Number of destinations that succeeded.
    pub destinations_succeeded: usize,
    /// Number of destinations that failed.
    pub destinations_failed: usize,
    /// Total duration in milliseconds.
    pub duration_ms: f64,
}

impl MultiWriteResult {
    /// Create a new result from destination results.
    pub fn from_results(results: Vec<DestinationResult>, total_duration: Duration) -> Self {
        let total_keys_written = results.iter().map(|r| r.keys_written).sum();
        let total_keys_failed = results.iter().map(|r| r.keys_failed).sum();
        let destinations_succeeded = results.iter().filter(|r| r.success).count();
        let destinations_failed = results.iter().filter(|r| !r.success).count();

        Self {
            results,
            total_keys_written,
            total_keys_failed,
            destinations_succeeded,
            destinations_failed,
            duration_ms: total_duration.as_secs_f64() * 1000.0,
        }
    }

    /// Check if all destinations succeeded.
    pub fn all_succeeded(&self) -> bool {
        self.destinations_failed == 0
    }
}

/// Writer for multiple Redis clusters/instances.
///
/// Executes writes to multiple destinations in parallel.
pub struct MultiClusterWriter {
    /// Destinations to write to.
    destinations: Vec<Destination>,
    /// Tokio runtime for async operations.
    runtime: Runtime,
}

impl MultiClusterWriter {
    /// Create a new multi-cluster writer.
    pub fn new(destinations: Vec<Destination>) -> Result<Self> {
        let runtime = Runtime::new()
            .map_err(|e| Error::Runtime(format!("Failed to create runtime: {}", e)))?;

        Ok(Self {
            destinations,
            runtime,
        })
    }

    /// Get the destinations.
    pub fn destinations(&self) -> &[Destination] {
        &self.destinations
    }

    /// Write hashes to all destinations in parallel.
    ///
    /// # Arguments
    /// * `keys` - Keys to write
    /// * `fields` - Field names
    /// * `values` - Values for each key (outer) and field (inner)
    /// * `ttl` - Optional TTL in seconds
    /// * `mode` - Write mode
    pub fn write_hashes(
        &self,
        keys: Vec<String>,
        fields: Vec<String>,
        values: Vec<Vec<Option<String>>>,
        ttl: Option<i64>,
        mode: WriteMode,
    ) -> Result<MultiWriteResult> {
        let start = Instant::now();

        let results: Vec<DestinationResult> = self.runtime.block_on(async {
            let futures: Vec<_> = self
                .destinations
                .iter()
                .map(|dest| {
                    let dest = dest.clone();
                    let keys = keys.clone();
                    let fields = fields.clone();
                    let values = values.clone();

                    async move {
                        let op_start = Instant::now();

                        // Apply key prefix
                        let prefixed_keys: Vec<String> =
                            keys.iter().map(|k| dest.prefix_key(k)).collect();

                        // Clone URL before moving dest
                        let url = dest.url.clone();
                        let timeout = dest.timeout_ms;

                        // Execute write with timeout
                        let result = tokio::time::timeout(
                            Duration::from_millis(timeout),
                            tokio::task::spawn_blocking(move || {
                                write_hashes(&url, prefixed_keys, fields, values, ttl, mode)
                            }),
                        )
                        .await;

                        let duration = op_start.elapsed();

                        match result {
                            Ok(Ok(Ok(write_result))) => {
                                DestinationResult::success(&dest, &write_result, duration)
                            },
                            Ok(Ok(Err(e))) => {
                                DestinationResult::failure(&dest, e.to_string(), duration)
                            },
                            Ok(Err(e)) => DestinationResult::failure(
                                &dest,
                                format!("Task panic: {}", e),
                                duration,
                            ),
                            Err(_) => DestinationResult::failure(
                                &dest,
                                format!("Timeout after {}ms", dest.timeout_ms),
                                duration,
                            ),
                        }
                    }
                })
                .collect();

            futures::future::join_all(futures).await
        });

        Ok(MultiWriteResult::from_results(results, start.elapsed()))
    }

    /// Write JSON documents to all destinations in parallel.
    ///
    /// # Arguments
    /// * `keys` - Keys to write
    /// * `documents` - JSON documents as strings
    /// * `ttl` - Optional TTL in seconds
    /// * `mode` - Write mode
    pub fn write_json(
        &self,
        keys: Vec<String>,
        documents: Vec<String>,
        ttl: Option<i64>,
        mode: WriteMode,
    ) -> Result<MultiWriteResult> {
        let start = Instant::now();

        let results: Vec<DestinationResult> = self.runtime.block_on(async {
            let futures: Vec<_> = self
                .destinations
                .iter()
                .map(|dest| {
                    let dest = dest.clone();
                    let keys = keys.clone();
                    let documents = documents.clone();

                    async move {
                        let op_start = Instant::now();

                        // Apply key prefix
                        let prefixed_keys: Vec<String> =
                            keys.iter().map(|k| dest.prefix_key(k)).collect();

                        // Clone URL before moving dest
                        let url = dest.url.clone();
                        let timeout = dest.timeout_ms;

                        // Execute write with timeout
                        let result = tokio::time::timeout(
                            Duration::from_millis(timeout),
                            tokio::task::spawn_blocking(move || {
                                write_json(&url, prefixed_keys, documents, ttl, mode)
                            }),
                        )
                        .await;

                        let duration = op_start.elapsed();

                        match result {
                            Ok(Ok(Ok(write_result))) => {
                                DestinationResult::success(&dest, &write_result, duration)
                            },
                            Ok(Ok(Err(e))) => {
                                DestinationResult::failure(&dest, e.to_string(), duration)
                            },
                            Ok(Err(e)) => DestinationResult::failure(
                                &dest,
                                format!("Task panic: {}", e),
                                duration,
                            ),
                            Err(_) => DestinationResult::failure(
                                &dest,
                                format!("Timeout after {}ms", dest.timeout_ms),
                                duration,
                            ),
                        }
                    }
                })
                .collect();

            futures::future::join_all(futures).await
        });

        Ok(MultiWriteResult::from_results(results, start.elapsed()))
    }
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_destination_new() {
        let dest = Destination::new("redis://localhost:6379");
        assert_eq!(dest.url, "redis://localhost:6379");
        assert!(dest.name.is_none());
        assert!(dest.key_prefix.is_empty());
        assert_eq!(dest.timeout_ms, 30000);
    }

    #[test]
    fn test_destination_builder() {
        let dest = Destination::new("redis://localhost:6379")
            .with_name("primary")
            .with_key_prefix("app:")
            .with_timeout_ms(5000);

        assert_eq!(dest.url, "redis://localhost:6379");
        assert_eq!(dest.name, Some("primary".to_string()));
        assert_eq!(dest.key_prefix, "app:");
        assert_eq!(dest.timeout_ms, 5000);
    }

    #[test]
    fn test_destination_display_name() {
        let dest1 = Destination::new("redis://localhost:6379");
        assert_eq!(dest1.display_name(), "redis://localhost:6379");

        let dest2 = Destination::new("redis://localhost:6379").with_name("primary");
        assert_eq!(dest2.display_name(), "primary");
    }

    #[test]
    fn test_destination_prefix_key() {
        let dest1 = Destination::new("redis://localhost:6379");
        assert_eq!(dest1.prefix_key("user:1"), "user:1");

        let dest2 = Destination::new("redis://localhost:6379").with_key_prefix("app:");
        assert_eq!(dest2.prefix_key("user:1"), "app:user:1");
    }

    #[test]
    fn test_destination_result_success() {
        let dest = Destination::new("redis://localhost:6379").with_name("test");
        let write_result = WriteResult {
            keys_written: 10,
            keys_failed: 0,
            keys_skipped: 0,
        };
        let duration = Duration::from_millis(100);

        let result = DestinationResult::success(&dest, &write_result, duration);

        assert_eq!(result.destination_name, "test");
        assert_eq!(result.keys_written, 10);
        assert_eq!(result.keys_failed, 0);
        assert!(result.success);
        assert!(result.error.is_none());
        assert!(result.duration_ms >= 100.0);
    }

    #[test]
    fn test_destination_result_failure() {
        let dest = Destination::new("redis://localhost:6379").with_name("test");
        let duration = Duration::from_millis(50);

        let result = DestinationResult::failure(&dest, "Connection refused".to_string(), duration);

        assert_eq!(result.destination_name, "test");
        assert_eq!(result.keys_written, 0);
        assert!(!result.success);
        assert_eq!(result.error, Some("Connection refused".to_string()));
    }

    #[test]
    fn test_multi_write_result() {
        let results = vec![
            DestinationResult {
                destination_name: "primary".to_string(),
                destination_url: "redis://host1:6379".to_string(),
                keys_written: 100,
                keys_failed: 5,
                success: true,
                error: None,
                duration_ms: 50.0,
            },
            DestinationResult {
                destination_name: "secondary".to_string(),
                destination_url: "redis://host2:6379".to_string(),
                keys_written: 0,
                keys_failed: 0,
                success: false,
                error: Some("Connection refused".to_string()),
                duration_ms: 10.0,
            },
        ];

        let multi_result = MultiWriteResult::from_results(results, Duration::from_millis(100));

        assert_eq!(multi_result.total_keys_written, 100);
        assert_eq!(multi_result.total_keys_failed, 5);
        assert_eq!(multi_result.destinations_succeeded, 1);
        assert_eq!(multi_result.destinations_failed, 1);
        assert!(!multi_result.all_succeeded());
    }

    #[test]
    fn test_multi_write_result_all_succeeded() {
        let results = vec![
            DestinationResult {
                destination_name: "primary".to_string(),
                destination_url: "redis://host1:6379".to_string(),
                keys_written: 100,
                keys_failed: 0,
                success: true,
                error: None,
                duration_ms: 50.0,
            },
            DestinationResult {
                destination_name: "secondary".to_string(),
                destination_url: "redis://host2:6379".to_string(),
                keys_written: 100,
                keys_failed: 0,
                success: true,
                error: None,
                duration_ms: 45.0,
            },
        ];

        let multi_result = MultiWriteResult::from_results(results, Duration::from_millis(60));

        assert_eq!(multi_result.total_keys_written, 200);
        assert_eq!(multi_result.destinations_succeeded, 2);
        assert_eq!(multi_result.destinations_failed, 0);
        assert!(multi_result.all_succeeded());
    }

    #[test]
    fn test_multi_cluster_writer_creation() {
        let destinations = vec![
            Destination::new("redis://localhost:6379").with_name("primary"),
            Destination::new("redis://localhost:6380").with_name("secondary"),
        ];

        let writer = MultiClusterWriter::new(destinations);
        assert!(writer.is_ok());

        let writer = writer.unwrap();
        assert_eq!(writer.destinations().len(), 2);
    }
}
